Source/eventHorizon/EventHorizons.ts

```diff
@@ -0,0 +1,110 @@
+    }
+
+    private throwIfSubscriptionDoesNotExist(subscription: Subscription) {
+        if (!this._subscriptionResponses.has(subscription)) {
+            throw new SubscriptionDoesNotExist(subscription);
+        }
+    }
+
+    private subscribeAll() {
+        for (const tenantWithSubscriptions of this.subscriptions) {
+            const consumerTenant = tenantWithSubscriptions.tenant;
+
+            for (const subscription of tenantWithSubscriptions.subscriptions) {
+                this._executionContextManager.currentFor(consumerTenant);
+
+                this._logger.debug(`Subscribing to events from ${subscription.partition} in ${subscription.stream} of ${subscription.tenant} in ${subscription.microservice} for ${consumerTenant} into ${subscription.scope}`);
+
+                const callContext = callContexts.toProtobuf(this._executionContextManager.current);
+                callContext.setHeadid(guids.toProtobuf(Guid.create()));
+
+                const pbSubscription = new PbSubscription();
+                pbSubscription.setCallcontext(callContext);
+                pbSubscription.setPartitionid(guids.toProtobuf(Guid.as(subscription.partition)));
+                pbSubscription.setScopeid(guids.toProtobuf(Guid.as(subscription.scope)));
+                pbSubscription.setStreamid(guids.toProtobuf(Guid.as(subscription.stream)));
+                pbSubscription.setTenantid(guids.toProtobuf(Guid.as(subscription.tenant)));
+                pbSubscription.setMicroserviceid(guids.toProtobuf(Guid.as(subscription.microservice)));
+
+                this._subscriptionsClient.subscribe(pbSubscription, (error: grpc.ServiceError | null, pbResponse?: PbSubscriptionResponse) => {
+                    const response = new SubscriptionResponse(guids.toSDK(pbResponse?.getConsentid()), failures.toSDK(pbResponse?.getFailure()));
+                    this._subscriptionResponses.set(subscription, response);
+
+                    this.handleResponse(tenantWithSubscriptions, subscription, response);
+
+                    if (response.failed) {
+                        this._logger.error(`Setting up event horizon subscription failed with '${response.failure?.reason}' - (id:${response.failure?.id}).`);
+                    }
+                });
+            }
+        }
+    }
+
+    private handleResponse(tenantWithSubscriptions: TenantWithSubscriptions, subscription: Subscription, response: SubscriptionResponse): void {
+        const consumerTenant = tenantWithSubscriptions.tenant;
+
+        subscription.handleResponse(consumerTenant, response);
+        tenantWithSubscriptions.handleResponse(consumerTenant, subscription, response);
+
+        this.completed(consumerTenant, subscription, response);
+
+        if (response.failed) {
+            this.failed(consumerTenant, subscription, response);
+        } else {
+            this.succeeded(consumerTenant, subscription, response);
+        }
```

Comment:
Since we loose type-safety when used in JS, should we check that these are actually functions before they are called? In that case we could also not set default empty functions on the top here.

Maybe we should use an event-emitter interface if there is such a thing in a nice TS package. That would probably deal with those things for us, and allow multiple callbacks if thats required.